File: packages/android/feedvalue/src/main/kotlin/com/feedvalue/sdk/FeedValueAPI.kt
package com.feedvalue.sdk

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.BufferedReader
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

// MARK: - API Response Types

internal data class ConfigResponse(
    val widgetId: String,
    val config: WidgetConfigData,
    val submissionToken: String?,
    val tokenExpiresAt: Long?
)

internal data class WidgetConfigData(
    val name: String,
    val promptText: String?,
    val thankYouMessage: String?
)

// MARK: - FeedValue API Client

internal class FeedValueAPI(baseUrl: String) {

    private val baseUrl = baseUrl.trimEnd('/')

    // Token lifecycle
    private var submissionToken: String? = null
    private var tokenExpiresAt: Long? = null
    private var fingerprint: String? = null

    // Cache
    private var cachedConfig: WidgetConfig? = null
    private var cacheExpiresAt: Long? = null
    private val cacheTTL: Long = 5 * 60 * 1000 // 5 minutes in milliseconds

    // Timeouts
    private val configTimeout = 10_000 // 10 seconds
    private val submitTimeout = 15_000 // 15 seconds

    fun setFingerprint(fingerprint: String) {
        this.fingerprint = fingerprint
    }

    // MARK: - Fetch Config

    suspend fun fetchConfig(widgetId: String, forceRefresh: Boolean = false): WidgetConfig {
        return withContext(Dispatchers.IO) {
            // Check cache
            if (!forceRefresh) {
                cachedConfig?.let { cached ->
                    cacheExpiresAt?.let { expires ->
                        if (System.currentTimeMillis() < expires) {
                            return@withContext cached
                        }
                    }
                }
            }

            val url = URL("$baseUrl/api/v1/widgets/$widgetId/config")
            val connection = url.openConnection() as HttpURLConnection

            try {
                connection.requestMethod = "GET"
                connection.connectTimeout = configTimeout
                connection.readTimeout = configTimeout

                fingerprint?.let {
                    connection.setRequestProperty("X-Client-Fingerprint", it)
                }
                connection.setRequestProperty("X-FeedValue-SDK", "android/1.0.0")

                val responseCode = connection.responseCode
                val responseBody = if (responseCode in 200..299) {
                    connection.inputStream.bufferedReader().use { it.readText() }
                } else {
                    connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
                }

                if (responseCode != 200) {
                    val message = parseErrorMessage(responseBody)
                    throw FeedValueError.ServerError(responseCode, message)
                }

                val configResponse = parseConfigResponse(responseBody)

                // Store token
                submissionToken = configResponse.submissionToken
                tokenExpiresAt = configResponse.tokenExpiresAt

                // Build widget config
                val widgetConfig = WidgetConfig(
                    widgetId = configResponse.widgetId,
                    name = configResponse.config.name,
                    promptText = configResponse.config.promptText,
                    thankYouMessage = configResponse.config.thankYouMessage
                )

                // Cache it
                cachedConfig = widgetConfig
                cacheExpiresAt = System.currentTimeMillis() + cacheTTL

                widgetConfig
            } catch (e: FeedValueError) {
                throw e
            } catch (e: Exception) {
                throw FeedValueError.NetworkError(e)
            } finally {
                connection.disconnect()
            }
        }
    }

    // MARK: - Submit Feedback

    suspend fun submitFeedback(
        widgetId: String,
        message: String,
        sentiment: Sentiment?,
        metadata: Map<String, String>
    ) {
        withContext(Dispatchers.IO) {
            // Refresh token if needed
            if (!hasValidToken()) {
                fetchConfig(widgetId, forceRefresh = true)
            }

            val token = submissionToken
                ?: throw FeedValueError.ServerError(403, "No submission token available")

            val url = URL("$baseUrl/api/v1/widgets/$widgetId/feedback")
            val connection = url.openConnection() as HttpURLConnection

            try {
                connection.requestMethod = "POST"
                connection.connectTimeout = submitTimeout
                connection.readTimeout = submitTimeout
                connection.doOutput = true

                connection.setRequestProperty("Content-Type", "application/json")
                connection.setRequestProperty("X-Submission-Token", token)
                fingerprint?.let {
                    connection.setRequestProperty("X-Client-Fingerprint", it)
                }
                connection.setRequestProperty("X-FeedValue-SDK", "android/1.0.0")

                // Build request body
                val body = JSONObject().apply {
                    put("message", message)
                    sentiment?.let { put("sentiment", it.value) }
                    if (metadata.isNotEmpty()) {
                        put("metadata", JSONObject(metadata))
                    }
                }

                OutputStreamWriter(connection.outputStream).use { writer ->
                    writer.write(body.toString())
                }

                val responseCode = connection.responseCode
                val responseBody = if (responseCode in 200..299) {
                    connection.inputStream.bufferedReader().use { it.readText() }
                } else {
                    connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
                }

                // Handle token refresh on 403
                if (responseCode == 403) {
                    val errorMessage = parseErrorMessage(responseBody)
                    if (errorMessage?.contains("token") == true || errorMessage?.contains("expired") == true) {
                        // Retry with fresh token
                        submissionToken = null
                        fetchConfig(widgetId, forceRefresh = true)

                        submissionToken?.let { newToken ->
                            retrySubmit(url, newToken, body)
                            return@withContext
                        }
                    }
                    throw FeedValueError.ServerError(responseCode, errorMessage)
                }

                if (responseCode !in 200..299) {
                    val errorMessage = parseErrorMessage(responseBody)
                    throw FeedValueError.ServerError(responseCode, errorMessage)
                }

                // Check for soft block
                try {
                    val json = JSONObject(responseBody)
                    if (json.optBoolean("blocked", false)) {
                        val blockMessage = json.optString("message", "Submission blocked")
                        throw FeedValueError.ServerError(403, blockMessage)
                    }
                } catch (e: Exception) {
                    // Ignore JSON parse errors for success response
                    if (e is FeedValueError) throw e
                }
            } catch (e: FeedValueError) {
                throw e
            } catch (e: Exception) {
                throw FeedValueError.NetworkError(e)
            } finally {
                connection.disconnect()
            }
        }
    }

    // MARK: - Helpers

    private fun hasValidToken(): Boolean {
        if (submissionToken == null) return false
        val expiresAt = tokenExpiresAt ?: return false
        // Add 30 second buffer before expiry
        val bufferMs = 30_000
        return System.currentTimeMillis() < (expiresAt * 1000 - bufferMs)
    }

    private fun retrySubmit(url: URL, token: String, body: JSONObject) {
        val connection = url.openConnection() as HttpURLConnection
        try {
            connection.requestMethod = "POST"
            connection.connectTimeout = submitTimeout
            connection.readTimeout = submitTimeout
            connection.doOutput = true

            connection.setRequestProperty("Content-Type", "application/json")
            connection.setRequestProperty("X-Submission-Token", token)
            fingerprint?.let {
                connection.setRequestProperty("X-Client-Fingerprint", it)
            }
            connection.setRequestProperty("X-FeedValue-SDK", "android/1.0.0")

            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(body.toString())
            }

            val responseCode = connection.responseCode
            if (responseCode !in 200..299) {
                val responseBody = connection.errorStream?.bufferedReader()?.use { it.readText() } ?: ""
                val errorMessage = parseErrorMessage(responseBody)
                throw FeedValueError.ServerError(responseCode, errorMessage)
            }
        } finally {
            connection.disconnect()
        }
    }

    private fun parseConfigResponse(json: String): ConfigResponse {
        val obj = JSONObject(json)
        val configObj = obj.getJSONObject("config")

        return ConfigResponse(
            widgetId = obj.getString("widget_id"),
            config = WidgetConfigData(
                name = configObj.getString("name"),
                promptText = configObj.optString("prompt_text", null),
                thankYouMessage = configObj.optString("thank_you_message", null)
            ),
            submissionToken = obj.optString("submission_token", null),
            tokenExpiresAt = if (obj.has("token_expires_at")) obj.getLong("token_expires_at") else null
        )
    }

    private fun parseErrorMessage(json: String): String? {
        return try {
            val obj = JSONObject(json)
            obj.optString("detail", null)
                ?: obj.optString("message", null)
                ?: obj.optString("error", null)
        } catch (e: Exception) {
            json.takeIf { it.isNotBlank() }
        }
    }
}

File: packages/ios/Sources/FeedValue/FeedValueAPI.swift
import Foundation

/// API response for config endpoint
struct ConfigResponse: Codable {
    let widgetId: String
    let config: WidgetConfigData
    let submissionToken: String?
    let tokenExpiresAt: Int?

    enum CodingKeys: String, CodingKey {
        case widgetId = "widget_id"
        case config
        case submissionToken = "submission_token"
        case tokenExpiresAt = "token_expires_at"
    }
}

struct WidgetConfigData: Codable {
    let name: String
    let promptText: String?
    let thankYouMessage: String?

    enum CodingKeys: String, CodingKey {
        case name
        case promptText = "prompt_text"
        case thankYouMessage = "thank_you_message"
    }
}

/// API response for feedback endpoint
struct FeedbackResponse: Codable {
    let feedbackId: String?
    let success: Bool?
    let blocked: Bool?
    let message: String?

    enum CodingKeys: String, CodingKey {
        case feedbackId = "feedback_id"
        case success
        case blocked
        case message
    }
}

/// Error response from API
struct ErrorResponse: Codable {
    let detail: String?
    let message: String?
    let error: String?
}

/// FeedValue API client
final class FeedValueAPI: @unchecked Sendable {

    private let baseUrl: String
    private let session: URLSession

    // Token lifecycle
    private var submissionToken: String?
    private var tokenExpiresAt: Int?
    private var fingerprint: String?

    // Cache
    private var cachedConfig: WidgetConfig?
    private var cacheExpiresAt: Date?
    private let cacheTTL: TimeInterval = 5 * 60 // 5 minutes

    // Timeouts
    private let configTimeout: TimeInterval = 10
    private let submitTimeout: TimeInterval = 15

    init(baseUrl: String) {
        self.baseUrl = baseUrl.trimmingCharacters(in: CharacterSet(charactersIn: "/"))

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = configTimeout
        self.session = URLSession(configuration: config)
    }

    func setFingerprint(_ fingerprint: String) {
        self.fingerprint = fingerprint
    }

    // MARK: - Fetch Config

    func fetchConfig(widgetId: String, forceRefresh: Bool = false) async throws -> WidgetConfig {
        // Check cache
        if !forceRefresh, let cached = cachedConfig, let expires = cacheExpiresAt, Date() < expires {
            return cached
        }

        guard let url = URL(string: "\(baseUrl)/api/v1/widgets/\(widgetId)/config") else {
            throw FeedValueError.serverError(statusCode: 0, message: "Invalid URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = configTimeout

        if let fingerprint = fingerprint {
            request.setValue(fingerprint, forHTTPHeaderField: "X-Client-Fingerprint")
        }
        request.setValue("ios/1.0.0", forHTTPHeaderField: "X-FeedValue-SDK")

        let (data, response) = try await performRequest(request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw FeedValueError.networkError(underlying: URLError(.badServerResponse))
        }

        if httpResponse.statusCode != 200 {
            let message = parseErrorMessage(from: data)
            throw FeedValueError.serverError(statusCode: httpResponse.statusCode, message: message)
        }

        let configResponse = try JSONDecoder().decode(ConfigResponse.self, from: data)

        // Store token
        self.submissionToken = configResponse.submissionToken
        self.tokenExpiresAt = configResponse.tokenExpiresAt

        // Build widget config
        let widgetConfig = WidgetConfig(
            widgetId: configResponse.widgetId,
            name: configResponse.config.name,
            promptText: configResponse.config.promptText,
            thankYouMessage: configResponse.config.thankYouMessage
        )

        // Cache it
        self.cachedConfig = widgetConfig
        self.cacheExpiresAt = Date().addingTimeInterval(cacheTTL)

        return widgetConfig
    }

    // MARK: - Submit Feedback

    func submitFeedback(
        widgetId: String,
        message: String,
        sentiment: Sentiment?,
        metadata: [String: String]
    ) async throws {
        // Refresh token if needed
        if !hasValidToken() {
            _ = try await fetchConfig(widgetId: widgetId, forceRefresh: true)
        }

        guard let token = submissionToken else {
            throw FeedValueError.serverError(statusCode: 403, message: "No submission token available")
        }

        guard let url = URL(string: "\(baseUrl)/api/v1/widgets/\(widgetId)/feedback") else {
            throw FeedValueError.serverError(statusCode: 0, message: "Invalid URL")
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = submitTimeout
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(token, forHTTPHeaderField: "X-Submission-Token")

        if let fingerprint = fingerprint {
            request.setValue(fingerprint, forHTTPHeaderField: "X-Client-Fingerprint")
        }
        request.setValue("ios/1.0.0", forHTTPHeaderField: "X-FeedValue-SDK")

        // Build request body
        var body: [String: Any] = ["message": message]

        if let sentiment = sentiment {
            body["sentiment"] = sentiment.rawValue
        }

        if !metadata.isEmpty {
            body["metadata"] = metadata
        }

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await performRequest(request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw FeedValueError.networkError(underlying: URLError(.badServerResponse))
        }

        // Handle token refresh on 403
        if httpResponse.statusCode == 403 {
            let errorMessage = parseErrorMessage(from: data)
            if errorMessage?.contains("token") == true || errorMessage?.contains("expired") == true {
                // Retry with fresh token
                self.submissionToken = nil
                _ = try await fetchConfig(widgetId: widgetId, forceRefresh: true)

                if let newToken = self.submissionToken {
                    request.setValue(newToken, forHTTPHeaderField: "X-Submission-Token")
                    let (retryData, retryResponse) = try await performRequest(request)

                    guard let retryHttpResponse = retryResponse as? HTTPURLResponse else {
                        throw FeedValueError.networkError(underlying: URLError(.badServerResponse))
                    }

                    if retryHttpResponse.statusCode >= 200 && retryHttpResponse.statusCode < 300 {
                        return // Success on retry
                    }

                    let retryMessage = parseErrorMessage(from: retryData)
                    throw FeedValueError.serverError(statusCode: retryHttpResponse.statusCode, message: retryMessage)
                }
            }
            throw FeedValueError.serverError(statusCode: httpResponse.statusCode, message: errorMessage)
        }

        if httpResponse.statusCode < 200 || httpResponse.statusCode >= 300 {
            let message = parseErrorMessage(from: data)
            throw FeedValueError.serverError(statusCode: httpResponse.statusCode, message: message)
        }

        // Check for soft block
        if let feedbackResponse = try? JSONDecoder().decode(FeedbackResponse.self, from: data),
           feedbackResponse.blocked == true {
            throw FeedValueError.serverError(statusCode: 403, message: feedbackResponse.message ?? "Submission blocked")
        }
    }

    // MARK: - Helpers

    private func hasValidToken() -> Bool {
        guard submissionToken != nil, let expiresAt = tokenExpiresAt else {
            return false
        }
        // Add 30 second buffer before expiry
        let bufferSeconds = 30
        return Int(Date().timeIntervalSince1970) < (expiresAt - bufferSeconds)
    }

    private func performRequest(_ request: URLRequest) async throws -> (Data, URLResponse) {
        do {
            return try await session.data(for: request)
        } catch let error as URLError {
            throw FeedValueError.networkError(underlying: error)
        } catch {
            throw FeedValueError.networkError(underlying: error)
        }
    }

    private func parseErrorMessage(from data: Data) -> String? {
        if let errorResponse = try? JSONDecoder().decode(ErrorResponse.self, from: data) {
            return errorResponse.detail ?? errorResponse.message ?? errorResponse.error
        }
        return String(data: data, encoding: .utf8)
    }
}
